File: apps/mobile-app/src/db/models/DevolucionPedido.ts
import { Model } from '@nozbe/watermelondb';
import { field, text, date, readonly, children } from '@nozbe/watermelondb/decorators';
import type { Query } from '@nozbe/watermelondb';
import type DetalleDevolucion from './DetalleDevolucion';

/**
 * Devolucion de un Pedido entregado. Parent de DetalleDevolucion (CASCADE).
 * TipoReembolso=SaldoFavor (0) genera credito al cliente; =Efectivo (1) sale del corte de caja.
 * Foto evidencia opcional via attachments (event_type='devolucion').
 * Sync push hacia backend SyncRepository.UpsertDevolucionAsync (parent + children atomico).
 */
export default class DevolucionPedido extends Model {
  static table = 'devoluciones_pedido';
  static associations = {
    detalle_devoluciones: { type: 'has_many' as const, foreignKey: 'devolucion_id' },
  };

  @field('server_id') serverId!: number | null;
  @field('pedido_id') pedidoId!: string;
  @field('pedido_server_id') pedidoServerId!: number | null;
  @field('cliente_id') clienteId!: string;
  @field('cliente_server_id') clienteServerId!: number | null;
  @field('usuario_id') usuarioId!: number;
  @field('ruta_id') rutaId!: string | null;
  @field('ruta_server_id') rutaServerId!: number | null;
  @date('fecha_devolucion') fechaDevolucion!: Date;
  @field('motivo') motivo!: number; // MotivoDevolucion enum
  @text('notas') notas!: string | null;
  @field('tipo_reembolso') tipoReembolso!: number; // 0=SaldoFavor, 1=Efectivo
  @field('monto_total') montoTotal!: number;
  @text('foto_evidencia_url') fotoEvidenciaUrl!: string | null;
  @field('estado') estado!: number; // 0=Activa, 1=Anulada
  @field('activo') activo!: boolean;
  @field('version') version!: number;
  @readonly @date('created_at') createdAt!: Date;
  @date('updated_at') updatedAt!: Date;

  @children('detalle_devoluciones') detalles!: Query<DetalleDevolucion>;
}

File: libs/HandySuites.Domain/Entities/DevolucionPedido.cs
using System.ComponentModel.DataAnnotations.Schema;
using HandySuites.Domain.Common;

namespace HandySuites.Domain.Entities;

public enum MotivoDevolucion
{
    DanoTransporte = 0,
    NoConforme = 1,
    FechaVencimiento = 2,
    ErrorPedido = 3,
    ClienteRetracta = 4,
    ProductoIncorrecto = 5,
    Otro = 99,
}

public enum TipoReembolso
{
    /// <summary>Genera saldo a favor del cliente (cliente.Saldo -= MontoTotal).</summary>
    SaldoFavor = 0,

    /// <summary>Vendedor reembolsa en efectivo: resta de su corte de caja.</summary>
    Efectivo = 1,
}

public enum EstadoDevolucion
{
    Activa = 0,
    Anulada = 1,
}

/// <summary>
/// Devolucion de productos de un Pedido entregado. Generada por el vendedor en mobile.
/// Side-effects al guardar:
/// - Si TipoReembolso=SaldoFavor: cliente.Saldo -= MontoTotal.
/// - Si TipoReembolso=Efectivo: resta de aRecibir en cierre de ruta.
/// - NO modifica Pedido.Total (venta original queda; devolucion es transaccion separada).
/// - NO incrementa inventario automatico (producto puede estar danado; reintegrar manual).
/// Validacion: solo permitida si pedido.estado=Entregado. Over-return validado por
/// sum(cantidades_devueltas_previas) por linea contra cantidad original del DetallePedido.
/// </summary>
[Table("DevolucionesPedido")]
public class DevolucionPedido : AuditableEntity
{
    [Column("id")]
    public int Id { get; set; }

    [Column("tenant_id")]
    public int TenantId { get; set; }

    [Column("mobile_record_id")]
    public string? MobileRecordId { get; set; }

    [Column("pedido_id")]
    public int PedidoId { get; set; }

    /// <summary>Denormalizado desde Pedido para queries rapidos (cartera por cliente).</summary>
    [Column("cliente_id")]
    public int ClienteId { get; set; }

    [Column("usuario_id")]
    public int UsuarioId { get; set; }

    [Column("ruta_id")]
    public int? RutaId { get; set; }

    [Column("fecha_devolucion")]
    public DateTime FechaDevolucion { get; set; } = DateTime.UtcNow;

    [Column("motivo")]
    public MotivoDevolucion Motivo { get; set; }

    [Column("notas")]
    public string? Notas { get; set; }

    [Column("tipo_reembolso")]
    public TipoReembolso TipoReembolso { get; set; } = TipoReembolso.SaldoFavor;

    [Column("monto_total")]
    public decimal MontoTotal { get; set; }

    /// <summary>URL de foto opcional (evidencia de dano). Stampeada via MobileAttachmentEndpoints.</summary>
    [Column("foto_evidencia_url")]
    public string? FotoEvidenciaUrl { get; set; }

    [Column("estado")]
    public EstadoDevolucion Estado { get; set; } = EstadoDevolucion.Activa;

    [Column("anulada_por")]
    public string? AnuladaPor { get; set; }

    [Column("anulada_en")]
    public DateTime? AnuladaEn { get; set; }

    /// <summary>Motivo de la anulacion (mirror de Gasto.MotivoInvalidacion). Capturado por supervisor.</summary>
    [Column("motivo_anulacion")]
    public string? MotivoAnulacion { get; set; }

    // Navigation properties
    public Tenant Tenant { get; set; } = null!;
    public Pedido Pedido { get; set; } = null!;
    public Cliente Cliente { get; set; } = null!;
    public Usuario Usuario { get; set; } = null!;
    public RutaVendedor? Ruta { get; set; }
    public ICollection<DetalleDevolucion> Detalles { get; set; } = new List<DetalleDevolucion>();
}
